util/chunk/mutrow.go

```diff
@@ -0,0 +1,338 @@
+		} else if len(rCol.offsets) > 0 {
+			setMutRowBytes(mrCol, rCol.data[rCol.offsets[row.idx]:rCol.offsets[row.idx+1]])
+		} else {
+			mrCol.ifaces[0] = rCol.ifaces[row.idx]
+		}
+		mrCol.nullBitmap[0] = 1
+	}
+}
+
+// SetValues sets the MutRow with values.
+func (mr MutRow) SetValues(vals ...interface{}) {
+	for i, v := range vals {
+		mr.SetValue(i, v)
+	}
+}
+
+// SetValue sets the MutRow with colIdx and value.
+func (mr MutRow) SetValue(colIdx int, val interface{}) {
+	col := mr.c.columns[colIdx]
+	if val == nil {
+		col.nullBitmap[0] = 0
+		return
+	}
+	switch x := val.(type) {
+	case int:
+		binary.LittleEndian.PutUint64(col.data, uint64(x))
+	case int64:
+		binary.LittleEndian.PutUint64(col.data, uint64(x))
+	case uint64:
+		binary.LittleEndian.PutUint64(col.data, x)
+	case float64:
+		binary.LittleEndian.PutUint64(col.data, math.Float64bits(x))
+	case float32:
+		binary.LittleEndian.PutUint32(col.data, math.Float32bits(x))
+	case string:
+		setMutRowBytes(col, hack.Slice(x))
+	case []byte:
+		setMutRowBytes(col, x)
+	case types.BinaryLiteral:
+		setMutRowBytes(col, x)
+	case types.Duration:
+		*(*types.Duration)(unsafe.Pointer(&col.data[0])) = x
+	case *types.MyDecimal:
+		*(*types.MyDecimal)(unsafe.Pointer(&col.data[0])) = *x
+	case types.Time:
+		col.ifaces[0] = x
+	case types.Enum:
+		setMutRowNameValue(col, x.Name, x.Value)
+	case types.Set:
+		setMutRowNameValue(col, x.Name, x.Value)
+	case json.JSON:
+		col.ifaces[0] = x
+	}
+	col.nullBitmap[0] = 1
+}
+
+// SetDatums sets the MutRow with datum slice.
+func (mr MutRow) SetDatums(datums ...types.Datum) {
+	for i, d := range datums {
+		mr.SetDatum(i, d)
+	}
+}
+
+// SetDatum sets the MutRow with colIdx and datum.
+func (mr MutRow) SetDatum(colIdx int, d types.Datum) {
+	col := mr.c.columns[colIdx]
+	if d.IsNull() {
+		col.nullBitmap[0] = 0
+		return
+	}
+	switch d.Kind() {
+	case types.KindInt64, types.KindUint64, types.KindFloat64:
+		binary.LittleEndian.PutUint64(mr.c.columns[colIdx].data, d.GetUint64())
+	case types.KindFloat32:
+		binary.LittleEndian.PutUint32(mr.c.columns[colIdx].data, math.Float32bits(d.GetFloat32()))
+	case types.KindString, types.KindBytes, types.KindBinaryLiteral:
+		setMutRowBytes(col, d.GetBytes())
+	case types.KindMysqlTime:
+		col.ifaces[0] = d.GetMysqlTime()
+	case types.KindMysqlDuration:
+		*(*types.Duration)(unsafe.Pointer(&col.data[0])) = d.GetMysqlDuration()
+	case types.KindMysqlDecimal:
+		*(*types.MyDecimal)(unsafe.Pointer(&col.data[0])) = *d.GetMysqlDecimal()
+	case types.KindMysqlJSON:
+		col.ifaces[0] = d.GetMysqlJSON()
+	case types.KindMysqlEnum:
+		e := d.GetMysqlEnum()
+		setMutRowNameValue(col, e.Name, e.Value)
+	case types.KindMysqlSet:
+		s := d.GetMysqlSet()
+		setMutRowNameValue(col, s.Name, s.Value)
+	default:
+		mr.c.columns[colIdx] = makeMutRowColumn(d.GetValue())
+	}
+	col.nullBitmap[0] = 1
+}
+
+func setMutRowBytes(col *column, bin []byte) {
+	if len(col.data) >= len(bin) {
+		col.data = col.data[:len(bin)]
```

Comment:
we may set like this?
``` go
col.data = col.data[:len(bin+1)]
col.nullBitmap = col.data[len(bin):]
```

Reply:
Then the data length is not the same as `len(bin)`.